Make button padding clickable and add screenshot compatibility

guillaumebreux opened the pull request:
Screenshots are zipped in the button component folder (i don't know how to do it otherwhise)

iRyusa commented:
Hi @guillaumebreux and thanks for your work ! 

I didn't checked your code yet, but can you provide a Litmus  with every buttons from this MJML here : 
https://github.com/mjmlio/mjml/pull/452#issuecomment-266486913

guillaumebreux commented:
Here it is: https://litmus.com/checklist/public/2878330

But there is some changes:
- I removed the `height` attribute which is not really relevent in my opinion.. (I can't make it work but if the wording are too long, it will break)
- I also removed the `inner-padding` since the padding is clickable now (or i dindn't understand the purpose of this attribute)
- I added a `box-shadow` attribute (almost the same browser support than `border-radius`, and i needed it for my own needs :p)

rogierslag commented:
OMG this is awesome!

![booyah](https://cloud.githubusercontent.com/assets/2778689/23022456/2d02fac0-f451-11e6-9552-024aab9bcd19.gif)

guillaumebreux commented on packages/mjml-button/src/index.js:
```diff
@@ -1,141 +1,190 @@
-const tagName = 'mj-button'
-const parentTag = ['mj-column', 'mj-hero-content']
-const endingTag = true
+const tagName               = 'mj-button';
```
Writing code without semi-colons is like writing a book without any punctuation :(
(But ye, i removed them for consistency)

iRyusa replied:
Well, I don't share this opinion. Some case still require `;` but it's doesn't add much to add semicolons, this will remain a debate in the community

guillaumebreux commented on packages/mjml-button/src/index.js:
```diff
@@ -1,141 +1,190 @@
     return helpers.merge({}, baseStyles, {
       table: {
-        width: mjAttribute('width')
+        margin:       mjAttribute('margin'),
```
Hmm i thought you would convert them with a capital M `margin:0;Margin:0` which is supported by outlook. But yes, let's use inner-padding instead

rogierslag replied:
Wait what!? At first I didnt believe it, but it seems to be true indeed https://www.emailonacid.com/blog/article/email-development/outlook.com-does-support-margins

guillaumebreux replied:
Thanks for the link @rogierslag 
I think mjml will still use padding for convenience though, adding margin now will add confusions imo. (But maybe it's a thing that can be debate around a PR/issue :D)

guillaumebreux commented:
Ok so I'm facing a weird issue right now, it is maybe something im not familiar with but:
The project add a `td` with margin where i have no control inside the button
```
  <mj-column>
    <mj-button padding="35px 10px">
  </mj-column>
```
is converted to:
```
  <table>
  <tbody>
    <tr>
      <td padding="35px 10px">
        <mj-button padding="35px 10px"> // the padding is still apply to where i want it inside the component
      </td>
    </tr>
  </tbody>
   </table>
```

And `<mj-column>` use this to insert component inside:
```
<tbody>
  {React.Children.map(children, child => React.cloneElement(child, { columnElement: true }))}
</tbody>
```

My component start with a root `<table>` so.. Where does this first `tr td` comes from? And why did they take my component's padding?

iRyusa commented:
This is because every elements inside `mj-column` has `tr / td` from MJMLElement decorator, that's why we had `inner-padding` & `padding`

iRyusa commented on packages/mjml-button/src/index.js:
```diff
@@ -51,85 +81,111 @@ class Button extends Component {
-          target="_blank" />
-      )
+        <a href={mjAttribute('href')} style={this.styles.a} target="_blank">
+          <table style={{borderCollapse: 'separate', width: '100%'}}>
```
Can this be moved inside baseStyle ?

iRyusa commented on packages/mjml-button/src/index.js:
```diff
@@ -51,85 +81,111 @@ class Button extends Component {
-              data-legacy-valign={mjAttribute('vertical-align')}
-              style={this.styles.td}>
-              {this.renderButton()}
+          <tr style={{padding: '0', textAlign: 'left', verticalAlign: 'top'}}>
```
Same here for consistency, please keep style inside baseStyle

guillaumebreux commented:
> This is because every elements inside mj-column has tr / td from MJMLElement decorator, that's why we had inner-padding & padding

So padding should be use for margin and inner-padding for the button padding ?

iRyusa commented:
Yep, it also works like that for every component like mj-image/mj-text/ and so on !

iRyusa commented:
Thanks for your hard work on this btw @GuillaumeBadi can you keep the litmus updated with last changes ?

guillaumebreux commented:
Sure here it is: https://litmus.com/checklist/public/a19599d
(you tagged the wrong person btw, but thanks :p)

guillaumebreux commented:
I fixed the width attribute that didn't work with percentage.
And in the meantimes, it fixes a UI issue you had with this where buttons could be wider than their outer box.
Before: ![before](https://cloud.githubusercontent.com/assets/8329354/23333715/b6da3722-fb90-11e6-993a-b0e6d9e3cb08.jpg)
Now: ![now](https://cloud.githubusercontent.com/assets/8329354/23333721/cb5a2572-fb90-11e6-8e7f-88bd30cb2ebb.jpg)

ztratar commented:
Guillaume -- thank you thank you thank you. When I recently discovered buttons weren't actually clickable aside from their text, my jaw dropped to the floor.

Looking forward for this code being merged in shortly to fix this issue. I think minor regressions in style on some email clients are still worth (what I consider) entirely non-functioning buttons.

guillaumebreux commented:
You're welcome @ztratar :)

guillaumebreux commented:
@iRyusa Any news about this? Do you need me to do more changes?

ngarnier commented:
hey @guillaumebreux, we're focusing the release of 3.3.0 to do things in order but will merge this shortly after!

iRyusa commented on packages/mjml-button/src/index.js:
```diff
@@ -51,85 +92,116 @@ class Button extends Component {
   getStyles () {
-    const { mjAttribute, defaultUnit } = this.props
+    const {mjAttribute} = this.props,
+      width = mjAttribute('width'),
```
Just for consistency can we remove multiline declaration ?

guillaumebreux replied:
Oh ok, is this considered a bad practice?

`Should we add a new attribute to allow current button behaviour ?`
Hmm, I don't think anyone would like the previous behaviour over the "normal" one. At worst, they won't care IMO.

iRyusa replied:
I'm not sure if this is a bad practice or not, but when adding more and more variable it becomes less and less readable.

I don't know, non clickable padding allow you to select the text inside a button more easily ex : http://codepen.io/anon/pen/peRBzW

guillaumebreux replied:
Hmm ye maybe i'll stop doing that for my own projects as well :D

You can start clicking from the outside of the button to select what's inside right?
Anyway it wont take too much time to code so.. your call :)

Fanetteg commented:
Hi, sorry I'm quite new to this.
Will this be implemented soon? I must say the unclickable padding + impossibility to add box-shadow is quite annoying :)

iRyusa commented:
We test this internally to make sure it will not break any major clients and keep same rendering as current mj-button. Visual regression for framework like MJML can be a major pain to migrate templates for users.

It should be good for 3.3.2, sorry about the delay !

Fanetteg commented:
Sweet. Thank you!

iRyusa commented:
Hi @guillaumebreux 

I'm merging this PR in its own branch to test it before shipping it. Huge thanks from MJML team and the community to the work on this, and I will ping you when it's ready to release 👍

startupseth commented:
My apologies if I'm missing something, from reading the PR, it looks like a solution was identified, and that it was released on Jun 8, 2017. Is this correct? I'm still unable to produce a button with clickable inner-padding in MJML today.

iRyusa commented:
This has been reverted because of an issue with all desktop outlook (buttons weren't clickable anymore with this patch)

skywinston commented:
@iRyusa I just fell in love with MJML using it for the first time today.  I found my way to this thread bc I also required a button with clickable inner-padding.  Since the last fix was reverted due to the issue, can you give us some guidance as to where we can track this issue (in another PR, perhaps)?

iRyusa commented:
Here's the issue https://github.com/mjmlio/mjml/issues/359 but there's no
really news on that.

On Thu, Mar 1, 2018 at 8:13 PM, Sky Winston <notifications@github.com>
wrote:

> @iRyusa <https://github.com/iryusa> I just fell in love with MJML using
> it for the first time today. I found my way to this thread bc I also
> required a button with clickable inner-padding. Since the last fix was
> reverted due to the issue, can you give us some guidance as to where we can
> track this issue (in another PR, perhaps)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mjmlio/mjml/pull/557#issuecomment-369698567>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAizzWStXGoogBGf_UdIOIgBktZ5ssHDks5taEhygaJpZM4MC5Gq>
> .
>



-- 
Cordialement,
*Maxime BRAZEILLES*

wiille commented:
Any news on this issue? Thanks for your work!

ngarnier commented:
well, as we don't have any better solution no, unfortunately there is no news.

wiille commented:
I found a ridiculous way to win some extra padding, adding html spaces before and after the text of the button. It's a ridiculous solution but it makes the button a bit bigger and works on all platforms:
```
<mj-button href="#">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; BUTTON TEXT &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</mj-button>
```

mikecreuzer commented:
I would simply argue that this component should be deleted. The **single** purpose of the component is to expand the area, so keeping it around is just misrepresenting the capabilities of this framework. Its a bit dramatic, but yeah its not capable of having a button component.

iRyusa commented:
@mikecreuzer  Purpose of this component is to stylize a bloc as button, we have numerous constraints to respect to have a fully clickable button that doesn't break when it's multiline, used within a background image, and have at least a decent fallback instead of being totally not clickable.

Here the component still allows you to style an element as a button without the overhead of creating a table each time manually. It generates like a dozen of HTML line for you.

Even big companies like Nike doesn't offer a fully clickable button on their marketing campaign, accessibility on every email clients is still a thing to have in mind.

I think you totally under estimate the work put behind such thing as a simple button.
